# Market Basket Analysis  - Affinity analysis - Association Rules

### Import necessary libraries

In [71]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

### Import Dataset

In [95]:
df_data = pd.read_csv('online_retail/data-year-2009-2010.csv', encoding="ISO-8859-1")

In [73]:
df_data

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

          InvoiceDate  Price  Customer ID         Country  
0        12/1/09 7:45   6.95      13085.0  United Kingdom  
1        12/1/09 7:45   6.75      13085.0  United Kingdom  
2        12/1/09 7:45   6.75      13085.0  United Kingdom  
3        12/1/09 7:45   2.10      13085.0  United Kingdom  
4        12/1/09 7:45   1.25      13085.0  United Kingdom  
...               ...    ...          ...             ...  
525456  12/9/10 20:01   2.95      17530.0  United Kingdom  
525457  12/9/10 20:01   3.75      17530.0  United Kingdom  
525458  12/9/10 20:01   3.75      17530.0  United Kingdom  
525459  12/9/10 20:01   3.75      17530.0  United Kingdom  
525460  12/9/10 20:01   1.95      17530.0  United Kingdom  

[525461 rows x 8 columns]

### Drop null values

In [74]:
df_data = df_data.dropna()

In [75]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417534 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      417534 non-null  object 
 1   StockCode    417534 non-null  object 
 2   Description  417534 non-null  object 
 3   Quantity     417534 non-null  int64  
 4   InvoiceDate  417534 non-null  object 
 5   Price        417534 non-null  float64
 6   Customer ID  417534 non-null  float64
 7   Country      417534 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 28.7+ MB


In [76]:
df_data['Country'].unique()

array(['United Kingdom', 'France', 'USA', 'Belgium', 'Australia', 'EIRE',
       'Germany', 'Portugal', 'Japan', 'Denmark', 'Netherlands', 'Poland',
       'Spain', 'Channel Islands', 'Italy', 'Cyprus', 'Greece', 'Norway',
       'Austria', 'Sweden', 'United Arab Emirates', 'Finland',
       'Switzerland', 'Unspecified', 'Nigeria', 'Malta', 'RSA',
       'Singapore', 'Bahrain', 'Thailand', 'Israel', 'Lithuania',
       'West Indies', 'Korea', 'Brazil', 'Canada', 'Iceland'],
      dtype=object)

In [77]:
df_data.describe()

Quantity          Price    Customer ID
count  417534.000000  417534.000000  417534.000000
mean       12.758815       3.887547   15360.645478
std       101.220424      71.131797    1680.811316
min     -9360.000000       0.000000   12346.000000
25%         2.000000       1.250000   13983.000000
50%         4.000000       1.950000   15311.000000
75%        12.000000       3.750000   16799.000000
max     19152.000000   25111.090000   18287.000000

In [78]:
df_data.describe(exclude='number')

Invoice StockCode                         Description   InvoiceDate  \
count   417534    417534                              417534        417534   
unique   23587      4031                                4459         21786   
top     500356    85123A  WHITE HANGING HEART T-LIGHT HOLDER  3/7/10 15:34   
freq       270      3245                                3245           270   

               Country  
count           417534  
unique              37  
top     United Kingdom  
freq            379423

In [79]:
df_data['InvoiceDate'] = pd.to_datetime(df_data['InvoiceDate'])

/var/folders/08/drbyjd9542qccyg42f0vybzh0000gn/T/ipykernel_5170/3972723281.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [80]:
print("""This Dataset start from {} 
                     to {}""".format(df_data['InvoiceDate'].describe()['first'],
                                     df_data['InvoiceDate'].describe()['last']))

This Dataset start from 2009-12-01 07:45:00 
                     to 2010-12-09 20:01:00


/var/folders/08/drbyjd9542qccyg42f0vybzh0000gn/T/ipykernel_5170/3884860247.py:2: FutureWarning:

Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.

/var/folders/08/drbyjd9542qccyg42f0vybzh0000gn/T/ipykernel_5170/3884860247.py:3: FutureWarning:

Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.



### Using the Positive ‘Quantity’ Values

In [81]:
data_plus = df_data[df_data['Quantity']>=0]
data_plus.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 407695 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      407695 non-null  object        
 1   StockCode    407695 non-null  object        
 2   Description  407695 non-null  object        
 3   Quantity     407695 non-null  int64         
 4   InvoiceDate  407695 non-null  datetime64[ns]
 5   Price        407695 non-null  float64       
 6   Customer ID  407695 non-null  float64       
 7   Country      407695 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 28.0+ MB


In [82]:
data_plus.describe()

Quantity          Price    Customer ID
count  407695.000000  407695.000000  407695.000000
mean       13.586686       3.294188   15368.504107
std        96.842229      34.756655    1679.795700
min         1.000000       0.000000   12346.000000
25%         2.000000       1.250000   13997.000000
50%         5.000000       1.950000   15321.000000
75%        12.000000       3.750000   16812.000000
max     19152.000000   10953.500000   18287.000000

In [83]:
top_10 = data_plus.groupby('Country').nunique().sort_values('Invoice', ascending=False).head(10)
top_10

Invoice  StockCode  Description  Quantity  InvoiceDate  Price  \
Country                                                                         
United Kingdom    17614       4001         4415       307        16601    366   
Germany             347       1457         1612        52          346     86   
EIRE                316       1999         2161        73          314    119   
France              236       1275         1398        94          235     82   
Netherlands         135        879          936        60          135     68   
Sweden               68        424          442        43           66     51   
Spain                66        682          702        35           66     61   
Belgium              52        481          521        24           52     55   
Switzerland          40        674          719        34           40     61   
Australia            40        418          451        37           40     49   

                Customer ID  
Country                      
United Kingdom         3971  
Germany                  67  
EIRE                      5  
France                   47  
Netherlands              22  
Sweden                   16  
Spain                    25  
Belgium                  17  
Switzerland              14  
Australia                15

In [84]:
top_10_transaction = pd.DataFrame(data_plus.groupby('Country').nunique().sort_values('Invoice', 
                                                                        ascending=False).head(10)['Invoice'])

In [85]:
top_10_transaction

Invoice
Country                
United Kingdom    17614
Germany             347
EIRE                316
France              236
Netherlands         135
Sweden               68
Spain                66
Belgium              52
Switzerland          40
Australia            40

In [86]:
# total bookings per market segment (incl. canceled)
segments=top_10_transaction

# pie plot
fig = px.pie(segments,
             values=top_10_transaction['Invoice'],
             names=top_10_transaction.index,
             title="Country Performance by Number of Invoice",
             template="seaborn")
fig.update_traces(rotation=-90, textinfo="percent+label")
fig.show()

### Create the basket data using the transaction from UK only

In [87]:
basket_plus = (data_plus[data_plus['Country'] =="United Kingdom"].groupby(['Invoice', 'Description'])['Quantity']
               .sum().unstack().reset_index().fillna(0).set_index('Invoice'))
basket_plus

Description    DOORMAT UNION JACK GUNS AND ROSES   3 STRIPEY MICE FELTCRAFT  \
Invoice                                                                       
489434                                       0.0                        0.0   
489435                                       0.0                        0.0   
489436                                       0.0                        0.0   
489437                                       0.0                        0.0   
489438                                       0.0                        0.0   
...                                          ...                        ...   
538167                                       0.0                        0.0   
538168                                       0.0                        0.0   
538169                                       0.0                        0.0   
538170                                       0.0                        0.0   
538171                                       0.0                        0.0   

Description   4 PURPLE FLOCK DINNER CANDLES   ANIMAL STICKERS  \
Invoice                                                         
489434                                  0.0               0.0   
489435                                  0.0               0.0   
489436                                  0.0               0.0   
489437                                  0.0               0.0   
489438                                  0.0               0.0   
...                                     ...               ...   
538167                                  0.0               0.0   
538168                                  0.0               0.0   
538169                                  0.0               0.0   
538170                                  0.0               0.0   
538171                                  0.0               0.0   

Description   BLACK PIRATE TREASURE CHEST   BROWN  PIRATE TREASURE CHEST   \
Invoice                                                                     
489434                                0.0                             0.0   
489435                                0.0                             0.0   
489436                                0.0                             0.0   
489437                                0.0                             0.0   
489438                                0.0                             0.0   
...                                   ...                             ...   
538167                                0.0                             0.0   
538168                                0.0                             0.0   
538169                                0.0                             0.0   
538170                                0.0                             0.0   
538171                                0.0                             0.0   

Description   Bank Charges   CAMPHOR WOOD PORTOBELLO MUSHROOM  \
Invoice                                                         
489434                 0.0                                0.0   
489435                 0.0                                0.0   
489436                 0.0                                0.0   
489437                 0.0                                0.0   
489438                 0.0                                0.0   
...                    ...                                ...   
538167                 0.0                                0.0   
538168                 0.0                                0.0   
538169                 0.0                                0.0   
538170                 0.0                                0.0   
538171                 0.0                                0.0   

Description   CHERRY BLOSSOM  DECORATIVE FLASK   FAIRY CAKE CANDLES  ...  \
Invoice                                                              ...   
489434                                     0.0                  0.0  ...   
489435                                     0.0    

In [88]:
basket_plus.tail()

Description    DOORMAT UNION JACK GUNS AND ROSES   3 STRIPEY MICE FELTCRAFT  \
Invoice                                                                       
538167                                       0.0                        0.0   
538168                                       0.0                        0.0   
538169                                       0.0                        0.0   
538170                                       0.0                        0.0   
538171                                       0.0                        0.0   

Description   4 PURPLE FLOCK DINNER CANDLES   ANIMAL STICKERS  \
Invoice                                                         
538167                                  0.0               0.0   
538168                                  0.0               0.0   
538169                                  0.0               0.0   
538170                                  0.0               0.0   
538171                                  0.0               0.0   

Description   BLACK PIRATE TREASURE CHEST   BROWN  PIRATE TREASURE CHEST   \
Invoice                                                                     
538167                                0.0                             0.0   
538168                                0.0                             0.0   
538169                                0.0                             0.0   
538170                                0.0                             0.0   
538171                                0.0                             0.0   

Description   Bank Charges   CAMPHOR WOOD PORTOBELLO MUSHROOM  \
Invoice                                                         
538167                 0.0                                0.0   
538168                 0.0                                0.0   
538169                 0.0                                0.0   
538170                 0.0                                0.0   
538171                 0.0                                0.0   

Description   CHERRY BLOSSOM  DECORATIVE FLASK   FAIRY CAKE CANDLES  ...  \
Invoice                                                              ...   
538167                                     0.0                  0.0  ...   
538168                                     0.0                  0.0  ...   
538169                                     0.0                  0.0  ...   
538170                                     0.0                  0.0  ...   
538171                                     0.0                  0.0  ...   

Description  ZINC HEART LATTICE CHARGER LARGE  \
Invoice                                         
538167                                    0.0   
538168                                    0.0   
538169                                    0.0   
538170                                    0.0   
538171                                    0.0   

Description  ZINC HEART LATTICE CHARGER SMALL  \
Invoice                                         
538167                                    0.0   
538168                                    0.0   
538169                                    0.0   
538170                                    0.0   
538171                                    0.0   

Description  ZINC HEART LATTICE DOUBLE PLANTER  \
Invoice                                          
538167                                     0.0   
538168                                     0.0   
538169                                     0.0   
538170                                     0.0   
538171                                     0.0   

Description  ZINC HEART LATTICE PLANTER BOWL  \
Invoice                                        
538167                                   0.0   
538168                                   0.0   
538169                                   0.0   
538170                                   0.0   
538171                                   0.0   

Description  ZINC HEART LATTICE T-LIGHT HOLDER  ZINC HEART LATTICE TRAY OVAL   \
Invo

### Encode The Data

In [89]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1
    
basket_encode_plus = basket_plus.applymap(encode_units)
basket_encode_plus

Description    DOORMAT UNION JACK GUNS AND ROSES   3 STRIPEY MICE FELTCRAFT  \
Invoice                                                                       
489434                                         0                          0   
489435                                         0                          0   
489436                                         0                          0   
489437                                         0                          0   
489438                                         0                          0   
...                                          ...                        ...   
538167                                         0                          0   
538168                                         0                          0   
538169                                         0                          0   
538170                                         0                          0   
538171                                         0                          0   

Description   4 PURPLE FLOCK DINNER CANDLES   ANIMAL STICKERS  \
Invoice                                                         
489434                                    0                 0   
489435                                    0                 0   
489436                                    0                 0   
489437                                    0                 0   
489438                                    0                 0   
...                                     ...               ...   
538167                                    0                 0   
538168                                    0                 0   
538169                                    0                 0   
538170                                    0                 0   
538171                                    0                 0   

Description   BLACK PIRATE TREASURE CHEST   BROWN  PIRATE TREASURE CHEST   \
Invoice                                                                     
489434                                  0                               0   
489435                                  0                               0   
489436                                  0                               0   
489437                                  0                               0   
489438                                  0                               0   
...                                   ...                             ...   
538167                                  0                               0   
538168                                  0                               0   
538169                                  0                               0   
538170                                  0                               0   
538171                                  0                               0   

Description   Bank Charges   CAMPHOR WOOD PORTOBELLO MUSHROOM  \
Invoice                                                         
489434                   0                                  0   
489435                   0                                  0   
489436                   0                                  0   
489437                   0                                  0   
489438                   0                                  0   
...                    ...                                ...   
538167                   0                                  0   
538168                   0                                  0   
538169                   0                                  0   
538170                   0                                  0   
538171                   0                                  0   

Description   CHERRY BLOSSOM  DECORATIVE FLASK   FAIRY CAKE CANDLES  ...  \
Invoice                                                              ...   
489434                                       0                    0  ...   
489435                                       0    

In [90]:
basket_encode_plus.tail()

Description    DOORMAT UNION JACK GUNS AND ROSES   3 STRIPEY MICE FELTCRAFT  \
Invoice                                                                       
538167                                         0                          0   
538168                                         0                          0   
538169                                         0                          0   
538170                                         0                          0   
538171                                         0                          0   

Description   4 PURPLE FLOCK DINNER CANDLES   ANIMAL STICKERS  \
Invoice                                                         
538167                                    0                 0   
538168                                    0                 0   
538169                                    0                 0   
538170                                    0                 0   
538171                                    0                 0   

Description   BLACK PIRATE TREASURE CHEST   BROWN  PIRATE TREASURE CHEST   \
Invoice                                                                     
538167                                  0                               0   
538168                                  0                               0   
538169                                  0                               0   
538170                                  0                               0   
538171                                  0                               0   

Description   Bank Charges   CAMPHOR WOOD PORTOBELLO MUSHROOM  \
Invoice                                                         
538167                   0                                  0   
538168                   0                                  0   
538169                   0                                  0   
538170                   0                                  0   
538171                   0                                  0   

Description   CHERRY BLOSSOM  DECORATIVE FLASK   FAIRY CAKE CANDLES  ...  \
Invoice                                                              ...   
538167                                       0                    0  ...   
538168                                       0                    0  ...   
538169                                       0                    0  ...   
538170                                       0                    0  ...   
538171                                       0                    0  ...   

Description  ZINC HEART LATTICE CHARGER LARGE  \
Invoice                                         
538167                                      0   
538168                                      0   
538169                                      0   
538170                                      0   
538171                                      0   

Description  ZINC HEART LATTICE CHARGER SMALL  \
Invoice                                         
538167                                      0   
538168                                      0   
538169                                      0   
538170                                      0   
538171                                      0   

Description  ZINC HEART LATTICE DOUBLE PLANTER  \
Invoice                                          
538167                                       0   
538168                                       0   
538169                                       0   
538170                                       0   
538171                                       0   

Description  ZINC HEART LATTICE PLANTER BOWL  \
Invoice                                        
538167                                     0   
538168                                     0   
538169                                     0   
538170                                     0   
538171                                     0   

Description  ZINC HEART LATTICE T-LIGHT HOLDER  ZINC HEART LATTICE TRAY OVAL   \
Invo

### Filter The Transaction : Bought More Than 1 Items Only

In [91]:
basket_filter_plus = basket_encode_plus[(basket_encode_plus > 0).sum(axis=1) >= 2]
basket_filter_plus

Description    DOORMAT UNION JACK GUNS AND ROSES   3 STRIPEY MICE FELTCRAFT  \
Invoice                                                                       
489434                                         0                          0   
489435                                         0                          0   
489436                                         0                          0   
489437                                         0                          0   
489438                                         0                          0   
...                                          ...                        ...   
538167                                         0                          0   
538168                                         0                          0   
538169                                         0                          0   
538170                                         0                          0   
538171                                         0                          0   

Description   4 PURPLE FLOCK DINNER CANDLES   ANIMAL STICKERS  \
Invoice                                                         
489434                                    0                 0   
489435                                    0                 0   
489436                                    0                 0   
489437                                    0                 0   
489438                                    0                 0   
...                                     ...               ...   
538167                                    0                 0   
538168                                    0                 0   
538169                                    0                 0   
538170                                    0                 0   
538171                                    0                 0   

Description   BLACK PIRATE TREASURE CHEST   BROWN  PIRATE TREASURE CHEST   \
Invoice                                                                     
489434                                  0                               0   
489435                                  0                               0   
489436                                  0                               0   
489437                                  0                               0   
489438                                  0                               0   
...                                   ...                             ...   
538167                                  0                               0   
538168                                  0                               0   
538169                                  0                               0   
538170                                  0                               0   
538171                                  0                               0   

Description   Bank Charges   CAMPHOR WOOD PORTOBELLO MUSHROOM  \
Invoice                                                         
489434                   0                                  0   
489435                   0                                  0   
489436                   0                                  0   
489437                   0                                  0   
489438                   0                                  0   
...                    ...                                ...   
538167                   0                                  0   
538168                   0                                  0   
538169                   0                                  0   
538170                   0                                  0   
538171                   0                                  0   

Description   CHERRY BLOSSOM  DECORATIVE FLASK   FAIRY CAKE CANDLES  ...  \
Invoice                                                              ...   
489434                                       0                    0  ...   
489435                                       0    

### Apply the Apriori Algorithm for the filtered data

In [92]:
frequent_itemsets_plus = apriori(basket_filter_plus, min_support=0.03, 
                                 use_colnames=True).sort_values('support', ascending=False).reset_index(drop=True)

frequent_itemsets_plus['length'] = frequent_itemsets_plus['itemsets'].apply(lambda x: len(x))

frequent_itemsets_plus

support                                        itemsets  length
0   0.179578            (WHITE HANGING HEART T-LIGHT HOLDER)       1
1   0.092402                      (REGENCY CAKESTAND 3 TIER)       1
2   0.077842                 (ASSORTED COLOUR BIRD ORNAMENT)       1
3   0.076784                (STRAWBERRY CERAMIC TRINKET BOX)       1
4   0.070500                      (HOME BUILDING BLOCK WORD)       1
..       ...                                             ...     ...
86  0.030490                          (POTTING SHED TEA MUG)       1
87  0.030303                (CREAM CUPID HEARTS COAT HANGER)       1
88  0.030241  (HEART OF WICKER LARGE, HEART OF WICKER SMALL)       2
89  0.030241                      (PLASTERS IN TIN SPACEBOY)       1
90  0.030116                (RED GINGHAM ROSE JEWELLERY BOX)       1

[91 rows x 3 columns]

In [93]:
frequent_itemsets_plus[ (frequent_itemsets_plus['length']  == 2) &
                        (frequent_itemsets_plus['support'] >= 0.03) ]

support                                           itemsets  length
41  0.042934  (RED HANGING HEART T-LIGHT HOLDER, WHITE HANGI...       2
62  0.036277  (SWEETHEART CERAMIC TRINKET BOX, STRAWBERRY CE...       2
71  0.033788  (WOODEN FRAME ANTIQUE WHITE , WOODEN PICTURE F...       2
82  0.031921  (HOME BUILDING BLOCK WORD, LOVE BUILDING BLOCK...       2
88  0.030241     (HEART OF WICKER LARGE, HEART OF WICKER SMALL)       2

### Finding The Association Between Frequently Bought Items

In [94]:
association_rules(frequent_itemsets_plus, metric='lift', 
                  min_threshold=1).sort_values('lift', ascending=False).reset_index(drop=True)

antecedents                           consequents  \
0         (WOODEN FRAME ANTIQUE WHITE )   (WOODEN PICTURE FRAME WHITE FINISH)   
1   (WOODEN PICTURE FRAME WHITE FINISH)         (WOODEN FRAME ANTIQUE WHITE )   
2      (SWEETHEART CERAMIC TRINKET BOX)      (STRAWBERRY CERAMIC TRINKET BOX)   
3      (STRAWBERRY CERAMIC TRINKET BOX)      (SWEETHEART CERAMIC TRINKET BOX)   
4               (HEART OF WICKER LARGE)               (HEART OF WICKER SMALL)   
5               (HEART OF WICKER SMALL)               (HEART OF WICKER LARGE)   
6            (HOME BUILDING BLOCK WORD)            (LOVE BUILDING BLOCK WORD)   
7            (LOVE BUILDING BLOCK WORD)            (HOME BUILDING BLOCK WORD)   
8    (RED HANGING HEART T-LIGHT HOLDER)  (WHITE HANGING HEART T-LIGHT HOLDER)   
9  (WHITE HANGING HEART T-LIGHT HOLDER)    (RED HANGING HEART T-LIGHT HOLDER)   

   antecedent support  consequent support   support  confidence       lift  \
0            0.060855            0.049095  0.033788    0.555215  11.309069   
1            0.049095            0.060855  0.033788    0.688213  11.309069   
2            0.047415            0.076784  0.036277    0.765092   9.964175   
3            0.076784            0.047415  0.036277    0.472447   9.964175   
4            0.061166            0.052828  0.030241    0.494405   9.358753   
5            0.052828            0.061166  0.030241    0.572438   9.358753   
6            0.070500            0.057682  0.031921    0.452780   7.849656   
7            0.057682            0.070500  0.031921    0.553398   7.849656   
8            0.058739            0.179578  0.042934    0.730932   4.070274   
9            0.179578            0.058739  0.042934    0.239085   4.070274   

   leverage  conviction  
0  0.030800    2.137898  
1  0.030800    3.012136  
2  0.032636    3.930114  
3  0.032636    1.805669  
4  0.027010    1.873380  
5  0.027010    2.195785  
6  0.027854    1.722011  
7  0.027854    2.081273  
8  0.032386    3.049127  
9  0.032386    1.237012

From the association_rules results above, we could see that WOODEN PICTURE FRAME WHITE FINISH and WOODEN FRAME ANTIQUE WHITE are the items that has the highest association each other since these two items has the highest “lift” value. The higher the lift value, the higher the association between the items willl. If the lift value is more than 1, it is enough for us to say that those two items are associated each other. In thise case, the highest value is 11.309069 which is very high. It means these 2 items are very good to be sold together.